### PR TITLE
🏗 Migrate `amphtml` package management from `yarn` to `npm`

### LINKED ISSUE #30518: I2I: Consider switching AMP package management from Yarn to NPM
## Background

AMP currently uses [`yarn`](https://yarnpkg.com/) for package management, and lists it as a requirement in the [developer docs](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md#one-time-setup). Yarn was [originally introduced](https://github.com/ampproject/amphtml/issues/2771) in 2016 because it provided a means for version pinning, which is critical to ensuring that the same environment is used by all developers and during CI.

Since then, NPM has introduced its own built-in mechanism for [version pinning](https://docs.npmjs.com/configuring-npm/package-locks.html). Initial efforts to eliminate the additional tool (yarn) were [abandoned](https://github.com/ampproject/amphtml/issues/11278#issuecomment-330571821) because npm's lockfile was found to be [inconsistent across OSs](https://github.com/npm/npm/issues/19966).

Meanwhile, Yarn's installation has become less reliable over time, with frequent failures on Travis and during local development (#15913, #23461). More recently, the latest version of yarn had issues like an invalid GPG signature and a version skew between github and npmjs.

## Actions

This issue is to study and track a potential switch back to NPM.

- [x] Investigate npm to see if past OS inconsistencies still exist
- [x] List the steps involved in switching back to npm
- [x] Understand the impact on the developer workflow (Will it just work? Or do devs need to uninstall stuff?)
- [x] Discuss this during design review and get approval from `wg-approvers`
- [x] Send out a PR to implement the switch
- [x] Co-ordinate PR changes with changes to AMP's internal release automation with help from `wg-infra`
- [ ] Merge all changes in the correct order

/cc @ampproject/wg-approvers @ampproject/wg-infra @kristoferbaxter

### PR DESCRIPTION
**Pull request by rsimha**
This PR changes AMP's package manager from `yarn` to `npm`. For a detailed discussion of why, see #30518.

**PR highlights:**
- Replaces all `yarn.lock` files in the repo with `package-lock.json`
- Updates Travis and GH Actions config files to no longer install / use / cache `yarn`
- Updates `check-package-manager.js` to error out when `yarn` is used to install dependencies
- Replaces the `yarn` integrity check for a stale `node_modules` with [`check-dependencies`](https://www.npmjs.com/package/check-dependencies), which runs in O(milliseconds)
- Updates PR check jobs to error out in the presence of uncommitted changes to `package-lock.json`
- Updates `validator/` scripts to use `npm install` instead of `yarn`
- Updates all `amphtml` developer docs

**Tested the following workflows:**
- Clean install (works)
- In-place install after syncing branch (works)
- Clean / in-place install when required tools are missing (both work, with expected warnings + steps)
- Install on different OSs (lockfile stable across Linux, macOS, and Windows)
- Run `gulp` task without installing packages (prompted to run `npm install`, `yarn` used to error out)
- Spot check changes between lockfiles (caret semver [resolution bug](https://github.com/ampproject/amphtml/pull/30700#issue-504619950) present in `yarn` has been fixed by `npm`)
- CI checks (sent out #30700 to pre-emptively fix some lint errors)

Fixes #30518

### CODE REVIEW

**Comment by rsimha** [comment]
This PR is now ready for preliminary review. Once reviewed and approved, I'll coordinate the merging of this PR with internal release system changes.

File: .eslintignore [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Output directories
-node_modules
+node_modules/**
```
Comment (rcebulko) [inline .eslintignore:2]: What is the difference here? I was under the impression a bare-name line in gitignore syntax ignored files/folders with that name (which implicitly includes all contained files and subfolders), and so the trailing `/**` here and elsewhere in this file shouldn't be necessary. Am I missing something?

File: build-system/common/check-package-manager.js [hunk 2]
```diff
@@ -63,34 +63,36 @@ function yellow(text) {
...
- * @fileoverview Perform checks on the AMP toolchain.
- */
-
-// If npm is being run, print a message and cause 'npm install' to fail.
-function ensureYarn() {
-  if (process.env.npm_execpath.indexOf('yarn') === -1) {
+ * If yarn is being run, print a message and cause 'yarn install' to fail.
+ **/
+function ensureNpm() {
+  if (process.env.npm_execpath.indexOf('npm') === -1) {
```
Comment (rcebulko) [inline build-system/common/check-package-manager.js:69]:
```suggestion
  if (process.env.npm_execpath.includes('npm')) {
```
Also nit: Maybe add a comment with a link to the documentation/StackOverflow thread discussing this env var

File: build-system/common/check-package-manager.js [hunk 3]
```diff
@@ -63,34 +63,36 @@ function yellow(text) {
...
     console.log(
-      red('*** The AMP project uses yarn for package management ***'),
+      red('*** The AMP project now uses npm for package management ***')
```
Comment (rcebulko) [inline build-system/common/check-package-manager.js:71]: Since this is one long constant string, I'd like to suggest making it a module-level constant:

```javascript
const NPM_INFO_MSG = `${red('*** The AMP project now uses npm for package management ***')}
For more info, see ${
  cyan('https://github.com/ampproject/amphtml/issues/30518')
}.

${yellow('To install all packages:')}
${cyan('$')} npm install

${yellow('To install a new (runtime) package to "dependencies":')}
${cyan('$')} npm install [package_name@version]

${yellow('To install a new (toolset) package to "devDependencies":')}
${cyan('$')} npm install --save-dev [package_name@version]

${yellow('To update a package:')}
${cyan('$')} npm update [package_name@version]

${yellow('To uninstall a package:')}
${cyan('$')} npm uninstall [package_name]

${yellow('For detailed instructions, see')} ${cyan(setupInstructionsUrl)}`
// ...
function ensureYarn() {
  if (process.env.npm_execpath.includes('npm')) {
    console.log(NPM_INFO_MSG);
  }
}
```

This might also make it more obvious to read/edit, since newlines and the like can be embedded directly in the same way it will be displayed

File: build-system/common/check-package-manager.js [hunk 4]
```diff
@@ -310,8 +327,8 @@ function checkPythonVersion() {
   // Python2 prints its version to stderr (fixed in Python 3.4)
   // See: https://bugs.python.org/issue18338
   const pythonVersionResult =
-    getStderr(`${pythonExecutable} --version`).trim() ||
-    getStdout(`${pythonExecutable} --version`).trim();
+    getStderr('python --version').trim() ||
+    getStdout('python --version').trim();
```
Comment (rcebulko) [inline build-system/common/check-package-manager.js:331]: `getStdout('python --version 2>&1').trim()` redirects stderr to stdout, rather than running this twice

File: build-system/common/check-package-manager.js [hunk 5]
```diff
@@ -349,17 +366,22 @@ function checkPythonVersion() {
...
+ * @return {Promise}
+ */
 async function main() {
-  // Yarn is already used by default on Travis, so there is nothing more to do.
-  if (process.env.TRAVIS) {
-    return 0;
+  // NPM is already used by default on Travis and Github Actions, so there is
+  // nothing more to do.
+  if (process.env.TRAVIS || process.env.GITHUB_ACTIONS) {
+    return Promise.resolve();
```
Comment (rcebulko) [inline build-system/common/check-package-manager.js:377]: Since it's an async function, `return;` is fine here (the result will be a Promise.resolve(undefined) anyway)

File: build-system/common/check-package-manager.js [hunk 6]
```diff
@@ -349,17 +366,22 @@ function checkPythonVersion() {
...
   }
-  ensureYarn();
+  ensureNpm();
   await checkNodeVersion();
   runGulpChecks();
   checkPythonVersion();
-  checkYarnVersion();
-  if (!process.env.TRAVIS && updatesNeeded.size > 0) {
+  logNpmVersion();
+  if (updatesNeeded.size > 0) {
```
Comment (rcebulko) [inline build-system/common/check-package-manager.js:384]:
```suggestion
  if (updatesNeeded.size) {
```

File: build-system/pr-check/npm-checks.js [hunk 7]
```diff
@@ -17,49 +17,46 @@
...
 const {gitDiffColor, gitDiffNameOnly} = require('../common/git');
 
 /**
- * Makes sure package.json and yarn.lock are in sync.
+ * Makes sure package.json and package-lock.json are in sync.
  * @param {string} fileName
  * @return {boolean}
  */
-function isYarnLockFileInSync(fileName = 'yarn-checks.js') {
+function isPackageLockFileInSync(fileName = 'npm-checks.js') {
```
Comment (rcebulko) [inline build-system/pr-check/npm-checks.js:33]: two nits:
- It doesn't look like there's any scenario in which it should be legal to call this without providing the filename (much like our `timedExec` helpers). It may be reasonable to drop this default altogether, given that it's unused and not intended to be used.
- If you're reluctant to require the file name, since it has a default, I think `opt_fileName` is preferred

File: contributing/TESTING.md [hunk 8]
```diff
@@ -410,7 +410,7 @@ You can create flamecharts for any node process used by the build system by leve
 Here's an example for `gulp dist --closure_concurrency=1`:
 
 ```sh
-yarn 0x -o node_modules/.bin/gulp dist --closure_concurrency=1
+npx 0x -o node_modules/.bin/gulp dist --closure_concurrency=1
 ```
 
-Important to node is `0x` will automatically create a flamechart and a serving folder locally within the repository, please don't add them to PRs!
+Important to note is `0x` will automatically create a flamechart and a serving folder locally within the repository, please don't add them to PRs!
```
Comment (rcebulko) [inline contributing/TESTING.md:416]: Not really part of this PR, but if the flamechart gets its own directory in the repo, we should probably add that to the gitignore

File: build-system/common/check-package-manager.js (hunk 4 above)
Comment (danielrozenberg) [reply build-system/common/check-package-manager.js:331]: Would using this shorthand break things on Windows?

File: build-system/common/check-package-manager.js [hunk 9]
```diff
@@ -279,27 +286,37 @@ function runGulpChecks() {
...
+    );
+  }
+}
+
+/**
+ * Checks if the local version of python is 2.7 or 3
+ */
 function checkPythonVersion() {
   // Python 2.7 is EOL but still supported
   // Python 3.5+ are still supported (TODO: deprecate 3.5 on 2020-09-13)
```
Comment (kristoferbaxter) [inline build-system/common/check-package-manager.js:322]: Looks like there is a note to deprecate 3.5 on 2020-09-13, worth looking into in a separate PR.

File: contributing/getting-started-e2e.md [hunk 10]
```diff
@@ -159,23 +159,15 @@ git branch -u upstream/master master
...
-- Install the stable version of [Yarn](https://yarnpkg.com/) (Mac and Linux: [here](https://yarnpkg.com/en/docs/install#alternatives-stable), Windows: [here](https://yarnpkg.com/lang/en/docs/install/#windows-stable))
-
-  ```sh
-  curl -o- -L https://yarnpkg.com/install.sh | bash
-  ```
-
-  An alternative to installing `yarn` is to invoke each Yarn command in this guide with `npx yarn` during local development. This will automatically use the current stable version of `yarn`.
-
-- Closure Compiler is automatically installed by Yarn, but it requires Java 8 which you need to install separately. AMP's version of Closure Compiler won't run with newer versions of Java. Download an installer for Mac, Linux or Windows [here](https://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html).
+- Closure Compiler is automatically installed by NPM, but it requires Java 8 which you need to install separately. AMP's version of Closure Compiler won't run with newer versions of Java. Download an installer for Mac, Linux or Windows [here](https://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html).
```
Comment (kristoferbaxter) [inline contributing/getting-started-e2e.md:170]: Might be able to omit this in a future version (we use the native build on Mac, Linux, and Windows now.

**Review by kristoferbaxter — APPROVED** [review]
A few minor comments, but overall looks great!

File: contributing/getting-started-e2e.md (hunk 10 above)
Comment (rsimha) [reply contributing/getting-started-e2e.md:170]: Also a good catch. There are other places where we can remove the need for Java. Will do in another follow up.

File: build-system/pr-check/npm-checks.js (hunk 7 above)
Comment (rsimha) [reply build-system/pr-check/npm-checks.js]: Nice catch. I've implemented your first suggestion.

**Review by rsimha — COMMENTED** [review]
Excellent comments all around! All addressed, except for the python deprecation and Java removal changes, which are out of scope here.

File: contributing/TESTING.md (hunk 8 above)
Comment (rsimha) [reply contributing/TESTING.md:416]: Nice idea. Tested this out. The flamechart output directory has a variable name and is generated in whichever directory the command is called. Can't really `.gitignore` it.

Comment (rcebulko) [reply contributing/TESTING.md:416]: I believe the directory is always `amphtml/\d{7}\.0x`, so `*.0x/` would do the trick

**Review by rcebulko — APPROVED** [review]
Updated comment re: flamegraph gitignore

File: contributing/TESTING.md (hunk 8 above)
Comment (rsimha) [reply contributing/TESTING.md:416]: Interesting idea. The [docs](https://github.com/davidmarkclements/0x#generating) say that the filename string is a PID, which can vary in shape and size across OSs. Given the already large scope of this PR, and the fact that flamechart creation is not part of the routine development workflow, I'm going to leave this as is for now.

**Comment by rsimha** [comment]
@honeybadgerdontcare for owners approval of changes in `validator/`
@jridgewell for owners approval of root-level `OWNERS` and `package-lock.json` files

**Review by honeybadgerdontcare — APPROVED** [review]

**Review by jridgewell — APPROVED** [review]

**Comment by kristoferbaxter** [comment]
`https://codelabs.developers.google.com/codelabs/creating-your-first-amp-component/index.html (404)` Looks like a failing URL. Checking master.

**Comment by rsimha** [comment]
> `https://codelabs.developers.google.com/codelabs/creating-your-first-amp-component/index.html (404)` Looks like a failing URL. Checking master.

We don't run the link-checker for all files, all the time on `master` because our docs contain thousands of links and it's virtually impossible to check them hundreds of times a day and get all green results. Therefore, we only re-check a documentation file when a PR modifies it.

I can fix this URL though. Would https://amp.dev/documentation/guides-and-tutorials/start/create work as a reasonable substitute?

**Review by kristoferbaxter — APPROVED** [review]

**Comment by rsimha** [comment]
Latest version of this PR has been verified on Linux, macOS, and Windows. Internal release code has been updated to work through the transition from yarn to npm. Merging this now 🤞.

**Comment by samouri** [comment]
Congrats on getting this through so quickly!  ⭐
